File: src/services/audit/detectors/ad/permissions/moderate/everyone-in-acl.ts
/**
 * EVERYONE_IN_ACL - Everyone/Authenticated Users with write permissions
 */

import { AclEntry } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { getUniqueObjects } from '../utils';

/**
 * Check for Everyone/Authenticated Users with write permissions
 */
export function detectEveryoneInAcl(aclEntries: AclEntry[], includeDetails: boolean): Finding {
  const EVERYONE_SID = 'S-1-1-0';
  const AUTHENTICATED_USERS_SID = 'S-1-5-11';
  const WRITE_MASK = 0x00020000; // ADS_RIGHT_DS_WRITE_PROP

  const affected = aclEntries.filter((ace) => {
    const isEveryone = ace.trustee === EVERYONE_SID || ace.trustee === AUTHENTICATED_USERS_SID;
    const hasWrite = (ace.accessMask & WRITE_MASK) !== 0;
    return isEveryone && hasWrite;
  });

  const uniqueObjects = getUniqueObjects(affected);
  const totalInstances = affected.length;

  return {
    type: 'EVERYONE_IN_ACL',
    severity: 'medium',
    category: 'permissions',
    title: 'Everyone in ACL',
    description: 'Everyone or Authenticated Users with write permissions in ACL. Overly permissive access.',
    count: uniqueObjects.length,
    totalInstances: totalInstances !== uniqueObjects.length ? totalInstances : undefined,
    affectedEntities: includeDetails ? uniqueObjects : undefined,
  };
}

File: internal/audit/detectors/ad/permissions/moderate/everyone-in-acl.go
package moderate

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// EveryoneInACLDetector detects Everyone/Authenticated Users with write permissions
type EveryoneInACLDetector struct {
	audit.BaseDetector
}

// NewEveryoneInACLDetector creates a new detector
func NewEveryoneInACLDetector() *EveryoneInACLDetector {
	return &EveryoneInACLDetector{
		BaseDetector: audit.NewBaseDetector("EVERYONE_IN_ACL", audit.CategoryPermissions),
	}
}

// Detect executes the detection
func (d *EveryoneInACLDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	const everyoneSID = "S-1-1-0"
	const authenticatedUsersSID = "S-1-5-11"
	const writeMask = 0x00020000 // ADS_RIGHT_DS_WRITE_PROP

	var affected []types.ACLEntry

	for _, ace := range data.ACLEntries {
		isEveryone := ace.Trustee == everyoneSID || ace.Trustee == authenticatedUsersSID
		hasWrite := (ace.AccessMask & writeMask) != 0

		if isEveryone && hasWrite {
			affected = append(affected, ace)
		}
	}

	uniqueObjects := helpers.GetUniqueObjects(affected)
	totalInstances := len(affected)

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Everyone in ACL",
		Description: "Everyone or Authenticated Users with write permissions in ACL. Overly permissive access.",
		Count:       len(uniqueObjects),
	}

	if totalInstances != len(uniqueObjects) {
		finding.TotalInstances = totalInstances
	}

	if data.IncludeDetails && len(uniqueObjects) > 0 {
		entities := make([]types.AffectedEntity, len(uniqueObjects))
		for i, dn := range uniqueObjects {
			entities[i] = types.AffectedEntity{
				Type: "object",
				DN:   dn,
			}
		}
		finding.AffectedEntities = entities
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewEveryoneInACLDetector())
}
